File: lib/python/opentoken/src/test/opentoken/attributes/general/integer_attribute_test.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import pytest
from opentoken.attributes.general.integer_attribute import IntegerAttribute


class TestIntegerAttribute:
    """Test cases for IntegerAttribute."""

    def setup_method(self):
        """Set up test fixtures before each test method."""
        self.attribute = IntegerAttribute()

    def test_get_name_should_return_integer(self):
        """Test that get_name returns 'Integer'."""
        assert self.attribute.get_name() == "Integer"

    def test_get_aliases_should_return_integer_alias(self):
        """Test that get_aliases returns list with 'Integer'."""
        aliases = self.attribute.get_aliases()
        assert len(aliases) == 1
        assert "Integer" in aliases

    def test_normalize_valid_integer_should_normalize(self):
        """Test normalization of valid integers."""
        assert self.attribute.normalize("123") == "123"
        assert self.attribute.normalize("0") == "0"
        assert self.attribute.normalize("-456") == "-456"
        assert self.attribute.normalize("  789  ") == "789"
        assert self.attribute.normalize("+123") == "123"

    def test_normalize_null_value_should_raise_error(self):
        """Test that None value raises ValueError."""
        with pytest.raises(ValueError):
            self.attribute.normalize(None)

    def test_normalize_invalid_integer_should_raise_error(self):
        """Test that invalid integers raise ValueError."""
        with pytest.raises(ValueError):
            self.attribute.normalize("abc")
        with pytest.raises(ValueError):
            self.attribute.normalize("12.34")
        with pytest.raises(ValueError):
            self.attribute.normalize("")

    def test_validate_valid_integer_should_return_true(self):
        """Test that valid integers validate successfully."""
        assert self.attribute.validate("123") is True
        assert self.attribute.validate("0") is True
        assert self.attribute.validate("-456") is True
        assert self.attribute.validate("  789  ") is True
        assert self.attribute.validate("+123") is True

    def test_validate_invalid_integer_should_return_false(self):
        """Test that invalid integers fail validation."""
        assert self.attribute.validate(None) is False
        assert self.attribute.validate("") is False
        assert self.attribute.validate("abc") is False
        assert self.attribute.validate("12.34") is False
        assert self.attribute.validate("1.0e10") is False

    def test_normalize_boundary_values_should_normalize(self):
        """Test normalization of boundary values."""
        assert self.attribute.normalize("0") == "0"
        assert self.attribute.normalize("-2147483648") == "-2147483648"
        assert self.attribute.normalize("2147483647") == "2147483647"

    def test_validate_large_integers_should_return_true(self):
        """Test that large integers validate successfully."""
        assert self.attribute.validate("9223372036854775807") is True
        assert self.attribute.validate("-9223372036854775808") is True

File: lib/java/opentoken/src/test/java/com/truveta/opentoken/attributes/general/IntegerAttributeTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.general;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class IntegerAttributeTest {
    private IntegerAttribute attribute;

    @BeforeEach
    void setUp() {
        attribute = new IntegerAttribute();
    }

    @Test
    void testGetName_ShouldReturnInteger() {
        assertEquals("Integer", attribute.getName());
    }

    @Test
    void testGetAliases_ShouldReturnIntegerAlias() {
        String[] aliases = attribute.getAliases();
        assertEquals(1, aliases.length);
        assertEquals("Integer", aliases[0]);
    }

    @Test
    void testNormalize_ValidInteger_ShouldReturnNormalized() {
        assertEquals("123", attribute.normalize("123"));
        assertEquals("0", attribute.normalize("0"));
        assertEquals("-456", attribute.normalize("-456"));
        assertEquals("789", attribute.normalize("  789  "));
    }

    @Test
    void testNormalize_WithPositiveSign_ShouldNormalize() {
        assertEquals("123", attribute.normalize("+123"));
    }

    @Test
    void testNormalize_NullValue_ShouldThrowException() {
        assertThrows(IllegalArgumentException.class, () -> attribute.normalize(null));
    }

    @Test
    void testNormalize_InvalidInteger_ShouldThrowException() {
        assertThrows(IllegalArgumentException.class, () -> attribute.normalize("abc"));
        assertThrows(IllegalArgumentException.class, () -> attribute.normalize("12.34"));
        assertThrows(IllegalArgumentException.class, () -> attribute.normalize(""));
    }

    @Test
    void testValidate_ValidInteger_ShouldReturnTrue() {
        assertTrue(attribute.validate("123"));
        assertTrue(attribute.validate("0"));
        assertTrue(attribute.validate("-456"));
        assertTrue(attribute.validate("  789  "));
        assertTrue(attribute.validate("+123"));
    }

    @Test
    void testValidate_InvalidInteger_ShouldReturnFalse() {
        assertFalse(attribute.validate(null));
        assertFalse(attribute.validate(""));
        assertFalse(attribute.validate("abc"));
        assertFalse(attribute.validate("12.34"));
        assertFalse(attribute.validate("1.0e10"));
    }

    @Test
    void testValidate_BoundaryValues_ShouldReturnTrue() {
        assertTrue(attribute.validate("0"));
        assertTrue(attribute.validate("-2147483648")); // Min int
        assertTrue(attribute.validate("2147483647")); // Max int
        assertTrue(attribute.validate("9223372036854775807")); // Max long
    }

}
